README.md
# <img src="https://raw.githubusercontent.com/remoshock/remoshock/master/src/remoshock/web/favicon.png" alt="Logo showing a drawing of a shock collar receiver" height="32">&nbsp;&nbsp; remoshock

Quick links:&nbsp;&nbsp;
<a href="https://remoshock.github.io/warning.html">
<img alt="Warning" src="https://img.shields.io/badge/-Warning-red"></a> 
<a href="https://remoshock.github.io/faq.html">
<img alt="Frequently Asked Questions" src="https://img.shields.io/badge/-FAQ-yellow"></a> 
<a href="https://remoshock.github.io/development/license.html">
<img alt="License: AGPL" src="https://img.shields.io/badge/-AGPL-%23AAF"></a> 
<a href="https://github.com/remoshock/remoshock/releases">
<img alt="Changes" src="https://img.shields.io/badge/-Changes-green"></a> 

<div>
<a href="https://remoshock.github.io/application/remoshockrnd.html">
<img style="border: 1px #AAA solid; padding: 1em" alt="Randomizer" src="https://raw.githubusercontent.com/remoshock/remoshock/master/docs/randomizer.png" height="200"></a> 
<a href="https://remoshock.github.io/receivers.html">
<img style="border: 1px #AAA solid; padding: 1em" alt="Photo of PAC collars with HackRF SDR" src="https://raw.githubusercontent.com/remoshock/remoshock/master/docs/pac.jpeg" height="200"></a> 
<a href="https://remoshock.github.io/application/remoshockserver.html">
<img style="border: 1px #AAA solid; padding: 1em" alt="Remote Control User Interface" src="https://raw.githubusercontent.com/remoshock/remoshock/master/docs/remote.png" height="200"></a>
</div>

remoshock is a computer based remote control for shock collars.

It consists of
- a web-based user interface, that works on mobile
- a randomizer program
- a command line interface program.
- a web-based API


Please see [https://remoshock.github.io/](https://remoshock.github.io/) for details.


<table>
<tr>

<tr><td>
<h2>🔧 Setup</h2>
<a href="https://remoshock.github.io/setup.html">
Hardware requirements, Getting started, Installations instructions.
</a>

</td><td>

<h2><img src="https://remoshock.github.io/resources/favicon_small.png" height="32" width="32" style="height: 1.2em; width: auto; vertical-align: text-bottom;" alt=""> Receivers</h2>
<a href="https://remoshock.github.io/receivers.html">
Supported shock collars: PAC, Petpat, Petrainer, Wodondog, ...
</a>

</td></tr>
<tr><td>

<h2>💻 Applications</h2>
<a href="https://remoshock.github.io/applications.html">
Mobile friendly web remote. Command line interface. Randomizer.
</a>

</td><td>

<h2>⌨️ Development</h2>
<a href="https://remoshock.github.io/development.html">
REST API, Source code.
</a>

</td></tr>
<tr><td>

<h2>❓ FAQ</h2>
<a href="https://remoshock.github.io/faq.html">
Frequently asked questons. Troubleshooting tips.
</a>

</td><td>

<h2>⚠️ Warning</h2>
<a href="https://remoshock.github.io/warning.html">
Playing with electricity has risks.
</a>

</td></tr>
</table>

docs/CHANGES.md
# Version 0.8.1

- **[gamepad]**    small improvements to the experimental game
- **[randomizer]** fixed input validation error in web randomizer


# Version 0.8 released on 2022-01-26

- **[randomizer]** web page for configuration as alternative to the existing command line tool
- **[web]**        improved layout of webapp
- **[web]**        require authentication for static pages as well
- **[gamepad]**    experimental support for games with gamepads


# Version 0.7 released on 2021-11-14

- **[receiver]** added support for Patpet T150 shock collars operating on 915 MHz
- **[receiver]** added support for second kind of Wodondog style shock collars
- **[rest]**     provide additional information about receivers via the REST API.
- **[internal]** dedicated action for keep-awake message to improve logging


# Version 0.6 released on 2021-10-12

- **[fix]**      improved reliability for PAC receivers by tuning off frequency to reduce
                 internal interferences in the HackRF hardware. 
                 This is a fix for a regression introduced in version 0.5.
- **[cli]**      action parameter is now case insensitive
- **[rest]**     added support for Bearer authentication to REST webservice
            

# Version 0.5 released on 2021-08-22
- **[package]**  made remoshock available on pip3 Python-repository
- **[package]**  changed name from pyshock to remoshock
- **[setup]**    on first run, a setup assistant helps with creating remoshock.ini
- **[receiver]** renamed "nameless" to "wodondog" receiver
- **[receiver]** added support for Petrainer


# Version 0.4 released on 2021-08-09

- **[receiver]** added support for a Nameless "Dog Shock Collar" depicted with a yellow remote control
- **[change]**   renamed pyshock.ini-setting button to channel
- **[internal]** internal URH invokation now supports ASK / OOK
- **[internal]** implemented a scheduler (e. g. for keep awake messages)
- **[internal]** automatically lint code and run tests on commits and pull requests


# Version 0.3 released on 2021-07-27

- **[randomizer]**   added start_delay_min_minutes and start_delay_max_minutes parameters
- **[randomizer]** / **[serve]** try to inhibit operating system hibernation, warning users if that was not possible
- **[doc]**         improved troubleshooting documentation
- **[development]** fixed linter warnings and minor bugs


# Version 0.2 released on 2021-07-24

- **[randomizer]**  replaced max_runtime_minutes with runtime_min_minutes and runtime_max_minutes
- **[core]**        suppress SDR debug message, unless --verbose is specified
- **[development]** fixed broken log messages in SDR-mock mode


# Version 0.1 released 2021-07-23

The very first release. Please report bugs, request features and give feedback on documentation and error messages.


pyshock/pacdog.py
#!/usr/bin/python3

import subprocess
from pyshock.pyshocklibdevices import Action, Device
from threading import RLock

lock = RLock()

class Pacdog(Device):

    button_codes = [
        #8 22 23
        [0, 0, 0],  # E/P left
        [0, 1, 1],  # B1  right 1
        [0, 1, 0],  # B2  right 2, E/P right
        [1, 1, 0],  # B3  right 3
        [1, 0, 0],  # B4  left 1
        [0, 0, 1],  # B5  left 2
        [1, 0, 1],  # B6  left 3
        [1, 1, 1]   # unused
    ]

    def __init__(self, name, color, code, button):
        super().__init__(name, color)
        self.code = code
        self.button = button

    def is_sdr_required(self):
        return True

    def boot(self, _arduino_manader, sdr_sender):
        self.sender = sdr_sender

    def generate(self, code, intensity, button, beep):
        pre_checksum = code[0:2] + self.calculate_intensity_code(intensity) + str(self.button_codes[button][0]) + code[2:]
        post_checksum = str(beep) + str(self.button_codes[button][1]) + str(self.button_codes[button][2])
        data = pre_checksum + "CCCCC" + post_checksum
        return pre_checksum + self.calculate_checksum(data) + post_checksum

    def calculate_intensity_code(self, intensity):
        res = ""
        for i in range(0, 6):
            res = res + str(intensity // 2**i % 2)
        return res

    def calculate_checksum(self, data):
        # a b c d e f g h i  j  k  l  m  n  o  p q   r  s
        # 7 6 5 4 3 2 1 0 15 14 13 12 11 10 09 8 23 22 21
        res =       str((int(data[0]) + int(data[ 8])) % 2)
        res = res + str((int(data[1]) + int(data[ 9]) + int(data[21])) % 2)
        res = res + str((int(data[2]) + int(data[10]) + int(data[22])) % 2)
        res = res + str((int(data[3]) + int(data[11]) + int(data[23])) % 2)
        res = res + str((int(data[4]) + int(data[12])) % 2)
        return res

    def encode(self, data):
        prefix = "0101010101010101111"
        filler = "10"
        res = prefix + filler
        for bit in data:
            res = res + bit + filler
        return res


    def send_cli(self, data):
        with lock:
            cmd = [
                "urh_cli",
                "--transmit",
                "--device", "HackRF",
                "--frequency", "27.1e6",
                "--sample-rate", "2e6",
                "--carrier-frequency", "27.1e6",
                "--modulation-type", "FSK",
                "--samples-per-symbol", "3100",
                "--parameters", "92e3", "95e3",
                "--pause", str(2 * 262924),
                # "--if-gain", "47",
                "--messages", data]
            print(cmd)
            subprocess.run(cmd)


    def send(self, data):
        with lock:
            samples = self.sender.modulate_messages(data)
            self.sender.send(samples)


    def command(self, action, level, duration):
        message = ""
        if action == Action.BEEPZAP:
            message = self.encode(self.generate(self.code, 0, self.button, 1)) + "/1s"

        beep = 1
        if action == Action.ZAP or action == Action.BEEPZAP:
            beep = 0

        if (duration < 250):
            duration = 250

        message_template = self.encode(self.generate(self.code, level * 63 // 100, self.button, beep))
        for _ in range(0, (duration + 5) // 250):
            message = message + " " + message_template

        self.send(message)


    def assertThat(self, message, expected, actual):
        if expected == actual:
            print(message + ": pass")
        else:
            print(message + ": FAILED")
            print("   Expected: " + expected)
            print("   Actual  : " + actual)

    def test_encoding(self):
        expected = "010101010101010111110010110010110010010110010010010110110010110110010010010010010010110110010"
        data = "010100100011011000000110"
        encoded = self.encode(data)
        self.assertThat("encoding", expected, encoded)

    def test_generate(self):
        expected = "010100100011011000000110"
        generated = self.generate("010110110", 18, 2, 1)
        self.assertThat("generation", expected, generated)

    def test_calculate_intensity_code(self):
        self.assertThat("intensity  1", "100000", self.calculate_intensity_code(1))
        self.assertThat("intensity 32", "000001", self.calculate_intensity_code(32))

    def test(self):
        self.test_encoding()
        self.test_generate()
        self.test_calculate_intensity_code()

"""
message = encode(generate("011100000", 20, 0, 1)) + "/1s"
for i in range(0, 4):
    message = message + " " + encode(generate("011100000", 20, 0, 0))

send(message)
"""

#test()

